### PR TITLE
TASK: Rename default authentication backend provider name

### PR DESCRIPTION
**Pull request by gerhard-boden**
Remove ``TYPO3`` reference from default authentication backend provider name and rename to ``Neos.Neos:Backend``.

#related https://github.com/neos/Neos.Demo/pull/23
#related https://github.com/Flowpack/Flowpack.Neos.FrontendLogin/pull/15

### CODE REVIEW

**Comment by gerhard-boden** [comment]
Please don't forget to also review 
https://github.com/neos/Neos.Demo/pull/23
https://github.com/Flowpack/Flowpack.Neos.FrontendLogin/pull/15

Thanks in advance for your valuable time guys :)

**Comment by gerhard-boden** [comment]
Tests are failing and I'm no really sure why:
https://travis-ci.org/neos/neos-development-collection/builds/192108094

I've renamed all the provider names in the settings. Maybe someone can help me out. ``Neos.Media.Browser`` is a separate package now, but still part of ``development-collection`` so it should be alright?

**Comment by bwaidelich** [comment]
Very nice!
I applied the three PRs locally and can confirm that this works. Tests are also green for me. They probably fail here because of the missing adjustments in `Neos.Demo` and/or `Flowpack.Neos.FrontendLogin`.

I just realized that we should probably use the chance to *namespace* the provider name like we did in the FE package.. (sorry, forgot about that when I suggested the name). WDYT?

**Comment by gerhard-boden** [comment]
Thanks a lot! Yea sure, shouldn't be too much trouble to add some prefix or suffix to the authenticator name. What is your suggestion for the final name?

**Comment by kdambekalns** [comment]
So, will this be adjusted to a namespaced identifier as suggested by @bwaidelich?

**Comment by gerhard-boden** [comment]
@kdambekalns yes, the suggested name by @bwaidelich was ``Neos.Neos:Default``. I will adjust this very soon.

**Comment by gerhard-boden** [comment]
Since I just wrote it: 
Wouldn't ``Neos.Neos:DefaultProvider`` be better when it comes to semantics?

**Comment by bwaidelich** [comment]
> Wouldn't Neos.Neos:DefaultProvider be better when it comes to semantics?

I don't think so. We also didn't call it `Flowpack.Neos.FrontendLogin:FrontendProvider` and usually never really add such suffix to the namespaced settings (like role identifiers, node type names, ...).

IMO `Neos.Neos:Default` is fine and the semantics become clear from the context

**Comment by gerhard-boden** [comment]
@bwaidelich : I updated the PRs to the new name.

The new name is okay with me, but I think it will be harder to migrate those identifiers again in case we want them changed, because `Neos.Neos:Default` could be used as a default in various places in the future. But I guess I'm thinking too much about stuff that might never happen ;)

Thanks again for reviewing!

File: Neos.Neos/Migrations/Code/Version20170115114620.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        return 'Neos.Neos-20170115114620';
+    }
+
+    /**
+     * @return void
+     */
+    public function up()
+    {
+        $this->searchAndReplace('Typo3BackendProvider', 'Neos.Neos:Default', ['php', 'ts2', 'fusion', 'js', 'json', 'html']);
+        $this->searchAndReplace('Typo3BackendProvider', "'Neos.Neos:Default'", ['yaml']);
```
Comment (bwaidelich) [inline Neos.Neos/Migrations/Code/Version20170115114620.php]: Previously you used `moveSettingsPaths` (which should take care of quoting itself).
Any particular reason why you changed that?

Comment (gerhard-boden) [inline Neos.Neos/Migrations/Code/Version20170115114620.php]: Yes, good that you mention it!
When I passed the path ``Neos.Flow.security.authentication.providers.Neos.Neos:Default`` it ended up in
```
providers:
  Neos:
    Neos:Default
```
I think the migration helper can't handle `Neos.Flow.security.authentication.providers[Neos.Neos:Default]`. Or is there a way to escape certain keys, so the settings path is no split up into seperate keys when a key contains `.`?

Comment (bwaidelich) [inline Neos.Neos/Migrations/Code/Version20170115114620.php]: I'll have a look

File: Neos.Neos/Migrations/Mysql/Version20170115114801.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
+namespace Neos\Flow\Persistence\Doctrine\Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration;
+use Doctrine\DBAL\Schema\Schema;
+
+/**
+ * Rename Typo3BackendProvider to Neos.Neos:Default
+ */
+class Version20170115114801 extends AbstractMigration
```
Comment (bwaidelich) [inline Neos.Neos/Migrations/Mysql/Version20170115114801.php]: Doctrine migrations should have a `getDescription` method (`./flow doctrine:migrationgenerate --diff-against-current false` will generate it)

File: Neos.Neos/Migrations/Code/Version20170115114620.php (hunk 1 above)
Comment (bwaidelich) [inline Neos.Neos/Migrations/Code/Version20170115114620.php]: OK, it seems we would need to implement that in `AbstractMigration::moveSettingsPaths()` and it won't be trivial probably..
So in this case I think it's safe to do it like you suggest

**Comment by gerhard-boden** [comment]
@bwaidelich thanks for the review, description methods have been added

**Comment by robertlemke** [comment]
I'm fine if we merge this with `Neos.Neos:Default`, but in my opinion this is too meaningless. It does have a meaning: it is the authentication provider for Neos backend users and not just some arbitrary default thing. Contrary to the `Neos.Flow:Default` provider, which is truly generic / defaultish.

Also personally I'd prefer having the "Provider" suffix since it's easier to get the meaning when you stumble over it in some debugging context.

Anyway, since I'm late to the discussion I'm fine with merging as is, but maybe you have some final thoughts? @gerhard-boden @bwaidelich

**Comment by gerhard-boden** [comment]
Thanks for your feedback @robertlemke!  In case you can persuade @bwaidelich to agree with the suffix, I would be fine with renaming the default provider once again :)

**Comment by kitsunet** [comment]
I guess the suffix can be helpful out of context (aka renamings and search/replace) so rather have it, even if it IS a bit duplicated...

**Comment by bwaidelich** [comment]
@robertlemke wrote:

> It does have a meaning: it is the authentication provider for Neos backend users

We could call it `Neos.Neos:DefaultBackend` or sth, but "backend" is still a somewhat clumsy term IMO. Besides (and more importantly): This is the standard provider to authenticate Neos users. It could be used in FE and for arbitrary API calls, too. The reason we use a different one for the FE Login example is that we don't _want_ to mix FE & BE Users for that

> Contrary to the Neos.Flow:Default provider [...]

Which `Neos.Flow:Default` provider are you referring to, AFAIK there is no such thing!?

> Also personally I'd prefer having the "Provider" suffix

But then we should adjust the [FE Login Provider](https://github.com/Flowpack/Flowpack.Neos.FrontendLogin/blob/master/Configuration/Settings.yaml#L11), too IMO.
Also, we currently don't use a suffix for lots of other namespaced tokens either (role identifiers, node types, ..) and usually the meaning gets quite clear from the context.

**Comment by robertlemke** [comment]
@bwaidelich I wouldn't use or suggest anyone to use the Neos Backend Provider (configuration) for frontend users. This kind of mixture caused all kinds of problems back then when I used TYPO3 …

I'd really like to have separate accounts (and thus authentication providers) for Backend, Frontend and API respectively.

I agree that "backend" is not an optimal term. But despite trying to establish an alternative, we always ended up calling it "backend" again, because that seems to be the term our users know best.

**Comment by bwaidelich** [comment]
@robertlemke Sorry, still not convinced ;)
The FE/BE mix in TYPO3 caused trouble mainly because of the separation it started with (`fe_users` vs `be_users` table) not because its a bad idea in general.
I often get the requirement that an editor should be able to log in to some FE area with his credentials..

Well, there are pros and cons with both and it's not really restricted by the default provider name anyways (it's just a configuration after all)..
But as you didn't respond to the other points: What would be your favorite name for it after the discussion here?

**Comment by bwaidelich** [comment]
Maybe `Neos.Neos:DefaultBackend` or `Neos.Neos:BackendDefault`?
I'd be OK with that. I'm certainly against the suffix ;)

**Comment by robertlemke** [comment]
> I often get the requirement that an editor should be able to log in to some FE area with his credentials..

It would be possible to automatically keep credentials of two accounts in sync. But if you only have 1 authentication provider for BE / FE you can't display the website for anonymous users while you are logged in as an editor. And that was the problem with TYPO3: as soon as you were logged in to the backend, the rendering of the frontend could change if it is tailored to making a difference between logged in and anonymous users.

As for the name, I'd like `Neos.Neos:BackendProvider` or `Neos.Neos:Backend` if we can't have the suffix. `DefaultBackend` is too confusing - which other backends exist apart from the default backend? (with the suffix it would still make sense: `DefaultBackendProvider`).

**Comment by bwaidelich** [comment]
> as soon as you were logged in to the backend, the rendering of the frontend could change [...]

That's what we have the `requestPattern` for, but nevermind, different topic really.

> As for the name, I'd like Neos.Neos:BackendProvider or Neos.Neos:Backend

I like `Neos.Neos:Backend` – and obviously I won't object if you all vote in favor of the `Provider` suffix. But keep in mind where you use this value:
* In the configuration underneath `authentication.providers`
* In the accounts table in the column `authenticationProviderName`
* In low-level PHP code such as the `$authenticationProviderName` argument of `AccountFactory::createAccountWithPassword()`

..all of which the meaning gets very obvious from the context IMO

**Comment by robertlemke** [comment]
@bwaidelich okay let's go for `Neos.Neos:Backend` then.

**Comment by robertlemke** [comment]
@gerhard-boden can you adjust your patch accordingly? Then we can merge it soon …

**Comment by gerhard-boden** [comment]
@robertlemke done (plus applied migrations again on neos/Neos.Demo#23 and Flowpack/Flowpack.Neos.FrontendLogin#15)

**Comment by aertmann** [comment]
they are.. `nodeFromLiveWorkspaceRemovedInPersonalWorkspaceExistsRemovedInGroupWorkspace` needs to be adjusted..

**Comment by robertlemke** [comment]
@gerhard-boden are you for fixing that last test? 🙏🏻

**Comment by gerhard-boden** [comment]
@aertmann @robertlemke The failing test is unrelated:
https://github.com/neos/neos-development-collection/blob/master/Neos.ContentRepository/Tests/Functional/Domain/LayeredWorkspacesTest.php#L143-L165

There is no relevant occurence of ``Typo3BackendProvider`` left, except for the two packages mentioned in the PR description. To me it seems rather coincidental that the Exception occurs while running this test. 

Can't reproduce locally either

**Comment by gerhard-boden** [comment]
Could it be that there are Settings somewhere within the test suite that need to be adjusted?

**Comment by gerhard-boden** [comment]
If nobody objects I would merge https://github.com/Flowpack/Flowpack.Neos.FrontendLogin/pull/15 now and restart the test afterwards

**Comment by gerhard-boden** [comment]
Thanks a lot to all of you for the vast amount of reviews! You guys rock! 🎸

**Comment by bwaidelich** [comment]
Thank _you_ for the vast amount of follow-up commits. You rock, too! 😀

**Comment by gerhard-boden** [comment]
No worries, it was mostly just search and replace ;)

**Pull request merged**